runtime/src/bank.rs

```diff
@@ -2734,9 +2734,10 @@ impl Bank {
 
     fn burn_and_purge_account(&self, program_id: &Pubkey, mut account: AccountSharedData) {
         self.capitalization.fetch_sub(account.lamports(), Relaxed);
-        // Resetting account balance to 0 is needed to really purge from AccountsDb and
-        // flush the Stakes cache
+        // Both resetting account balance to 0 and zeroing the account data
+        // is needed to really purge from AccountsDb and flush the Stakes cache
         account.set_lamports(0);
+        account.data_as_mut_slice().fill(0);
```

Comment:
@ryoqun can you take a look at this? When the `stakes_remove_delegation_if_inactive` feature was activated, we no longer remove delegations simply if the account balance is zero for stake accounts. The data must still be large enough to hold stake state AND it must not deserialize into a state that has a delegation.

Reply:
This change will only affect consensus if we create a new builtin for an active stake account.

Reply:
haha, you're editing scary part. thanks for pinging :)

to be sure, account with lamport=0 should always be same account hash regardless of its data contents. that's why this doesn't need feature-gated, right?

i mean, there's risk in `add_builtin_account` (https://github.com/solana-labs/solana/blob/1149c1880df89190fdfbc5a18859531c371d5f39/runtime/src/bank.rs#L2814) (likewise, in `add_precompiled_account`) under mixed-version cluster condition _unless the above premise holds true_.

Reply:
Yes, exactly, the account hash shouldn't change. The risk in builtin/precompiled accounts is that the stakes cache state is corrupted so maybe worth adding a new feature gate in your opinion? There is a new change (https://github.com/solana-labs/solana/pull/21654) that prunes corrupt state from the stakes cache but I'd rather not test it.

Reply:
spoiler:

![image](https://user-images.githubusercontent.com/117807/145827742-54aba3c2-7abe-49bd-bb47-371f3c4038af.png)

Reply:
> The risk in builtin/precompiled accounts is that the stakes cache state is corrupted so maybe worth adding a new feature gate in your opinion?

i don't think it's worth adding a new feature gate. i always forget the following impracticality of these attacks... xD:

https://github.com/solana-labs/solana/blob/1149c1880df89190fdfbc5a18859531c371d5f39/runtime/src/bank.rs#L2807-L2814

Reply:
```rust
// Both resetting account balance to 0 and zeroing the account data
// is needed to really purge from AccountsDb and flush the Stakes cache
account.set_lamports(0);
account.data_as_mut_slice().fill(0);
```
What does above comment actually mean?
I see zero lamports accounts are not stored in accounts-db even if they have non zero data and data_len.
Also, the code indicates that is the case:
https://github.com/solana-labs/solana/blob/64abd008c/runtime/src/accounts_db.rs#L4971-L4978

This causes stakes cache to be inconsistent with accounts-db. Whereas previously if lamports = 0, the delegation was removed from stakes cache as well.

Reply:
This method is only called when we forcefully purge an account so this checks ensures that if we purge an active stake account, it *actually* gets removed from the stakes cache. Simply setting the lamports to 0 is not enough because the stake account state could still be deserialized into a delegation and then `update_stake_delegation` would be called with `Some(0, Delegation)` and `self.stake_delegations.remove(stake_pubkey)` wouldn't be called.

Reply:
I think the code comment could be better expressed, because as it is right now, it may falsely imply that setting account balance to 0 may not always remove the account from _AccountsDb_, which to my understanding it is not true.

But that issue aside and ignoring accounts-db for the moment, regarding the stakes-cache:
1. old code used to remove delegations if `lamports == 0`: https://github.com/solana-labs/solana/blob/476124de5/runtime/src/stakes.rs#L181-L185
2. not removing delegations from stakes-cache when `lamports == 0`  will cause stakes cache to be inconsistent with accounts-db. ie. there will be zombie delegations in stakes cache with no associated stake account in accounts-db. 

Is there any scenario that we **don't** want to remove the delegation from the stakes-cache when account's `lamports == 0`?